Adds support for signing in with multiple identities. Closes #3587

martinlingstuyl opened the pull request:
Closes #3587

🚀 There we go again! Adding multiple connection functionality 🚀

- Added logic to allow to login to multiple identities
- Added a command to list available connections
- Added a command to switch to another connection
- Added a command to update a connection name
- Added a command to remove a connection
- Added an option to assign a connection name when signing in.
- Added a message about what identity you're using when running in debug mode.
- Added `connectionName` to the output of commands like `m365 status` and `m365 login`.
- The `connection use` output is the same as `m365 status`

**Points of attention**

- We need to test if Managed Identity still works. That's the only one I haven't tested yet.

- __No graceful fallback__ I've changed the shape of the data that's saved to our token caches. As a result I've decided (with Waldek) to just use different filenames. So after installing everyone needs to run m365 login again. 

Specifically I've added:

Name | Explanation
--|--
`name` | The name of the connection. By default this is the localAccountId of the associated account or the objectId of the Entra principal.  (so unique, also in b2b scenario's)
`identityName` | The name of the identity, can be a userName or application name 
`identityId` | The localAccountId of the associated account or just the objectId of the AAD principal. Is necessary to locate the identity and find it in the MSAL store.

**Notes for reviewers**

- I've overridden the base action on the `connection <verb>` command. The reason is that we need to be able to execute these commands when we don't have an active connection. So for example: If I have two connections and I remove one (which is also my active connection) I'll be left in a semi-logged-out state. I need to be able to execute `connection use` to switch to another connection, hence I'm overriding the base command action.

- There's some changes in test cases that seem to be unrelated to the PR. For example: `spo page add`, `spo page set`, `spo site ensure`, `todo task add`, `todo task set`. These changes are due to issues that were invisible till now. (missing stubs that cause issues with authentication for example)

waldekmastykarz commented:
Hey @martinlingstuyl, is this PR right? Seems like we've got some extra files modified that aren't a part of the proposed change. Could you please have another look?

martinlingstuyl commented:
Can you name the ones you mean? I know I had issues with some tests, which caused me to do more changes than what seems necessary

martinlingstuyl commented:
Also: do check out the PR description where I mention a few of those

waldekmastykarz commented:
> Can you name the ones you mean? I know I had issues with some tests, which caused me to do more changes than what seems necessary

I'm seeing 777 files changed and included in the PR, a bunch of them are docs. Are all of those changes necessary? Also, it seems like we've got some conflicts. 😊

waldekmastykarz commented:
> > Can you name the ones you mean? I know I had issues with some tests, which caused me to do more changes than what seems necessary
> 
> I'm seeing 777 files changed and included in the PR, a bunch of them are docs. Are all of those changes necessary? Also, it seems like we've got some conflicts. 😊

My bad, the files seem to be modified tests. All good. If you could please resolve conflicts, then I'll start the review.

martinlingstuyl commented:
done @waldekmastykarz !

waldekmastykarz commented on src/m365/connection/commands/connection-use.spec.ts:
```diff
@@ -0,0 +1,188 @@
+    assert.notStrictEqual(command.description, null);
+  });
+
+  it('fails validation if the id is not a valid guid', async () => {
```
`connection use` doesn't have an `id` option

waldekmastykarz commented on src/m365/connection/commands/connection-use.spec.ts:
```diff
@@ -0,0 +1,188 @@
+  });
+
+  it(`switches to the 'Contoso Application' identity using the name option`, async () => {
+    await assert.doesNotReject(command.action(logger, { options: { name: 'acd6df42-10a9-4315-8928-53334f1c9d01' } }));
```
Let's use a single assert per test to simplify our tests. What is it that we actually want to verify. Let's update other tests as well.

waldekmastykarz commented on src/m365/connection/commands/connection-use.ts:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  public get description(): string {
+    return "When signed in with multiple identities, switch to another connection";
```
Let's use single quotes

waldekmastykarz replied:
Since the command can also be used with one connection, let's simplify its name to `Activate the specified Microsoft 365 tenant connection`

waldekmastykarz commented on src/m365/connection/commands/connection-set.ts:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  public get description(): string {
+    return "When signed in with multiple identities, update a specified connection";
```
Let's simplify the name to `Rename the specified connection`

waldekmastykarz commented on src/m365/connection/commands/connection-remove.ts:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  public get description(): string {
+    return "When signed in with multiple identities, switch to another connection";
```
Let's simplify the name to `Remove the specified connection`

waldekmastykarz reviewed (changes requested):
Very nice! Let's do a couple of changes before we merge.

In general, let's pay attention to the following:
- let's use single quotes where possible. If we need to escape an apostrophe, let's use backticks
- let's use a single assert per test. Otherwise, if a test fails, the only way to find out what's wrong is to debug it, which isn't convenient
- let's avoid to use `!` to suppress null validation. It can lead to runtime errors, which we could otherwise catch design time thanks to typings

martinlingstuyl commented on src/m365/connection/commands/connection-remove.ts:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  public get description(): string {
+    return "When signed in with multiple identities, switch to another connection";
```
You'd think I learn :-)

waldekmastykarz replied:
It's consistently inconsistent 😄

martinlingstuyl commented on src/m365/connection/commands/connection-remove.spec.ts:
```diff
@@ -0,0 +1,157 @@
+
+    sinon.stub(auth, 'ensureAccessToken').resolves();
+
+    auth.connection.active = true;
```
I don't think so, as it will remove connections and change the active connection state in a couple of runs. Better to just have it here to keep from repeating myself.

martinlingstuyl replied:
You're correct after all. I'm not updating the array. I'm now just checking the correct auth method is called.

martinlingstuyl commented on src/Auth.ts:
```diff
@@ -190,22 +225,24 @@ export class Auth {
+    if (this.connection.authType !== AuthType.Certificate
+      && this.connection.authType !== AuthType.Secret
+      && this.connection.authType !== AuthType.Identity) {
+      this.clientApplication = await this.getPublicClient(logger, debug);
```
Correct, if you look at the original code, you'll see that the condition was only implemented for cert auth. But secret and identity auth are very similar and would simply be skipped somewhat later, when retrieving accounts. See below code. I just made it more explicit that these auth methods are excluded from the entire `ensureAccessTokenSilent` part.

```ts
if (this.service.authType !== AuthType.Certificate) {
  this.clientApplication = await this.getClientApplication(logger, debug);
  if (this.clientApplication) {
    const accounts = await this.clientApplication.getTokenCache().getAllAccounts();
    // Secret and Identity would be exluded here, as they have no account as well.
    if (accounts.length > 0) {
      getTokenPromise = this.ensureAccessTokenSilent.bind(this);
    }
  }
}
```

waldekmastykarz replied:
Looking at the code, I might've mistaken `!==` for `===`. Good point.

martinlingstuyl commented on src/Auth.ts:
```diff
@@ -391,10 +417,11 @@ export class Auth {
+
     return (this.clientApplication as Msal.ClientApplication).acquireTokenSilent({
-      account: accounts[0],
+      account: account!,
```
We've already checked earlier that an account exists. (When setting the `ensureAccessTokenSilent` promise) I'm actually not changing the code here from what it was. it used to be `accounts[0]`, which may be `undefined` as well. I'm not sure if we should add a condition here, as that would just lead to extra test cases for a situation that should not occur.

martinlingstuyl replied:
I could of course throw an exception here if account is undefined. But my gut sense says I'd want to exclude it from the code coverage.

waldekmastykarz replied:
The danger here is that we're making assumptions based on the current flow/implementation. Should anything change, we'll miss this and get runtime exceptions. If the code above, where we retrieve the account can retrieve null/undefined, we should check for it, just to be safe.

martinlingstuyl replied:
The account should never be null, as this code is only reached through the following:

```ts
if (accounts.filter(a => a.localAccountId === this.connection.identityId).length > 0 && this.connection.active === true) {
          getTokenPromise = this.ensureAccessTokenSilent.bind(this);
}
```

But OK, I'll just throw an exception if the account is empty.

martinlingstuyl commented on src/Auth.ts:
```diff
@@ -713,34 +740,86 @@ export class Auth {
+
+      if (this.clientApplication) {
+        const tokenCache = this.clientApplication.getTokenCache();
+        const account = await tokenCache.getAccountByLocalId(connection!.identityId!);
```
I've removed one !. 
What's your suggestion for the last one? `identityId` is nullable. I'm expecting it to be filled here, but I dislike having to do a null-check at every fricking place where you use such a thing.

how would you advise to solve it?

waldekmastykarz replied:
If it's nullable, why are you expecting it to be filled here? Why is that assumption implicit? It's the whole point of using types to isolate yourself from implicit assumptions. If something is optional, we should check for it before using it. If the value is always present, then we shouldn't make it optional.

martinlingstuyl replied:
I'm expecting it to be filled, because what we're doing here is removing a connection. We wouldn't be here if there is no connection, a connection always has an identityId value after signing in. Hence it's logical to assume identityId is filled.

It would be better if the identityId property of the connection object would not be nullable. But that would mean we'd need to refactor the code again. The original service object/interface setup and auth flow was created in a less strict typescript era I think 😆

martinlingstuyl replied:
I've now worked around it as you can see in my last commit. As I'm saying: this is not exactly 100% ideal. But I do not wish to refactor the auth.connection object setup further.

martinlingstuyl commented on src/Auth.ts:
```diff
@@ -260,22 +302,6 @@ export class Auth {
     return response.accessToken;
   }
 
-  private async getClientApplication(logger: Logger, debug: boolean): Promise<Msal.ClientApplication | undefined> {
```
This method was only used twice, and in one of those case I've excluded all application auth types because that made more sense. Using just getPublicClient and getConfidentialClient started to made more sense then.

```ts
if (this.connection.authType !== AuthType.Certificate &&
      this.connection.authType !== AuthType.Secret &&
      this.connection.authType !== AuthType.Identity) {
      this.clientApplication = await this.getPublicClient(logger, debug);
//...
}
```

martinlingstuyl commented on src/Auth.ts:
```diff
@@ -750,6 +829,63 @@ export class Auth {
+
+  public getIdentityDetails(identity: Connection, debug: boolean): IdentityDetails {
+    const details: IdentityDetails = {
+      connectionName: identity.name!,
```
because the connection object is already created before the identityId and name values are available. (Before signing in)

waldekmastykarz replied:
If the values could possibly be empty, then we should check for them rather than assume they're not empty.

martinlingstuyl replied:
Same problem as the one above... Let's discuss it there first.

martinlingstuyl replied:
I've now worked around it as you can see in my last commit. As I'm saying: this is not exactly 100% ideal. But I do not wish to refactor the auth.connection object setup further.

martinlingstuyl commented on src/m365/commands/identityDetails.ts:
```diff
@@ -0,0 +1,9 @@
+  authType: string;
+  appId: string;
+  appTenant: string;
+  accessTokens?: string;
```
Good catch! It ought to be plural. I'll move it to command level, so I can distinguish between them.

martinlingstuyl commented:
Hi @waldekmastykarz, it's kinda hard to find comment replies on GH is my experience, (when there are a lot of comments) if you reply to a resolved comment, could you unresolved it?

waldekmastykarz commented on src/Auth.ts:
```diff
@@ -750,6 +837,63 @@ export class Auth {
+      throw new CommandError(`The connection name '${newName}' is already in use`);
+    }
+
+    allConnections.forEach(c => {
```
To avoid iterating over all connections twice, let's update the code, so that we store the outcome of allConnections.filter in an array. We can then use it to compare its length first, and if we found only one object, directly update its name.

martinlingstuyl replied:
I've changed it a bit, as there's only one entry of the connection available.

waldekmastykarz commented on src/Auth.ts:
```diff
@@ -750,6 +837,63 @@ export class Auth {
+    const allConnections = await this.getAllConnections();
+    const connections = allConnections.filter(i => i.name === name);
+
+    if (!connections || connections.length === 0) {
```
If we only care about the first instance, then let's use `allConnections.find` instead of `.filter`. `.find` immediately returns the first matching object or undefined if no such object found.

martinlingstuyl commented:
Were those all comments you had @waldekmastykarz ? Or is the review not done yet?

waldekmastykarz commented:
> Were those all comments you had @waldekmastykarz ? Or is the review not done yet?

This was it. You've done awesome job 👏  Let me have another look and we'll get it in for folks to test.

martinlingstuyl commented:
I'll see if the conditionally nullable and assert does a better job instead of the null checks

martinlingstuyl commented:
Ok @waldekmastykarz, I fixed your comments. I also added some `assert` lines instead of if-statements. See the latest commit. The most interesting thing happened: login tests started failing, because the code was run, while these properties were not available. I corrected that in the login tests.

waldekmastykarz commented:
Just to double check: is the PR ready for review?

martinlingstuyl commented:
Yes it is! Unless I must first fix the merge conflicts 😀

martinlingstuyl commented:
All yours @waldekmastykarz!

waldekmastykarz reviewed (changes requested):
I've just noticed an undesired side-effect: when you run `m365 status` and you're not signed in, you get a login prompt. Let's revert this behavior to what we have now (status shows the current state without automatically requiring you to login).

martinlingstuyl commented:
Yeah, that's definitely not as it should. I'll check it out and update it

martinlingstuyl commented:
I cannot reproduce this behavior @waldekmastykarz, 
I've logged out from all accounts. Running `m365 status` now shows me `"Logged out"`
Now I've signed in with two accounts and removed the active one using `m365 connection remove`. Running the status command now shows me `"Logged out, signed in connections available"`. Running m365 logout again, now the status command again shows me `"Logged out"`.

Can you consistently reproduce it in some way?

waldekmastykarz commented:
> I cannot reproduce this behavior @waldekmastykarz, I've logged out from all accounts. Running `m365 status` now shows me `"Logged out"` Now I've signed in with two accounts and removed the active one using `m365 connection remove`. Running the status command now shows me `"Logged out, signed in connections available"`. Running m365 logout again, now the status command again shows me `"Logged out"`.
> 
> Can you consistently reproduce it in some way?

Odd, I can't repro it indeed. All seems to be working just fine. I wonder if it's related to the fact, that I was signed in with the current code base and then pulled in your PR, built it and ran the status command. Either way, all seems good. I'll proceed with the review. Thanks for double checking and sorry for the trouble.

waldekmastykarz commented:
Merged manually. Brilliant work! So cool to see this out there 👏

martinlingstuyl commented:
This is really a game changer indeed. Superb!
